### PR TITLE
Use EnvironmentFile instead of Environment

### PR DESCRIPTION
**Pull request by rariyama**
## Overview
fix https://github.com/treasure-data/digdag/issues/1749

I made a change that [environment file](https://docs.aws.amazon.com/AmazonECS/latest/APIReference/API_ContainerOverride.html#:~:text=Required%3A%20No-,environmentFiles,-A%20list%20of) is used when passing the values of environment to the container instead of [environment](https://docs.aws.amazon.com/AmazonECS/latest/APIReference/API_ContainerOverride.html#:~:text=environment,or%20the%20task%20definition.%20You%20must%20also%20specify%20a%20container%20name.), that is the way to add them to the container one by one. 
This change will be effective to avoid failing an execution of ECS Run Task because [a maximum size of characters of ECS container overrides is 8192](https://docs.aws.amazon.com/AmazonECS/latest/APIReference/API_RunTask.html#:~:text=A%20list%20of,Type%3A%20TaskOverride%20object) and this limit includes environments defined by the current implementation.
it will be helpful especially when executing `sh>:` operator. In addition to variables defined in `_env`, variables defined in `_export` are also passed to the container unlike `py>:` and `rb>:` so the size of environment tends to be larger as the development progresses.

## The change points
- to upgrade awsJavaSdkVersion
    - it needs to use [EnvironmentFiles](https://docs.aws.amazon.com/AWSJavaSDK/latest/javadoc/com/amazonaws/services/ecs/model/EnvironmentFile.html)
- to use [withEnvironmentFiles](https://docs.aws.amazon.com/AWSJavaSDK/latest/javadoc/com/amazonaws/services/ecs/model/ContainerOverride.html#withEnvironmentFiles-com.amazonaws.services.ecs.model.EnvironmentFile...-:~:text=the%20container%20definition.-,withEnvironmentFiles,-public%C2%A0ContainerOverride) instead of [withEnvironment](https://docs.aws.amazon.com/AWSJavaSDK/latest/javadoc/com/amazonaws/services/ecs/model/ContainerOverride.html#withEnvironmentFiles-com.amazonaws.services.ecs.model.EnvironmentFile...-:~:text=a%20container%20name.-,withEnvironment,-public%C2%A0ContainerOverride)

### CODE REVIEW

**Comment by rariyama** [comment]
I added the parameter for choosing the existing way in this commit https://github.com/treasure-data/digdag/pull/1762/commits/27d1ae3340fc00afa7b8e21bc6a6c62f8c31a268.
Users can select the either way for passing environments to a container by setting it. The default value is false, so there is no impact for existing users. 
If some users want to use environment file on most of tasks in workflow, they can use it easily by setting the parameter on the root task and override it on child ones as necessary.

**Comment by yoyama** [comment]
Sorry for very late reply.
I don't check it in detail, but I have a few concerns.

- I think providing an option to use file to pass environment variables is good and default is false is also good.
  But I think it is better to be configured by server properties because it is internal behavior.
- Your change increases the code line of the method. I hope refactoring on your change.

**Comment by rariyama** [comment]
@yoyama 
Thanks for your reply.

> I think providing an option to use file to pass environment variables is good and default is false is also good.
But I think it is better to be configured by server properties because it is internal behavior.

It makes sense to me, so I modified my implementation with this commit https://github.com/treasure-data/digdag/pull/1762/commits/fe6f8bfca83ee15eaeee01dbff5b437562e9a7b9. I left `setEcsContainerOverrideEnvironment` and added a process about using environmentFiles to it.

> Your change increases the code line of the method. I hope refactoring on your change.

I also dealt with this concern with the commit above. I extracted a process about uploading environmentFile to S3 as another method from `setEcsContainerOverrideEnvironment`.

If there are any other concerns, please let me know about those.

File: digdag-standards/src/main/java/io/digdag/standards/command/ecs/TemporalProjectArchiveStorage.java [hunk 1]
```diff
@@ -55,4 +55,11 @@ public InputStream getContentInputStream(final String key)
             throw ThrowablesUtil.propagate(e);
         }
     }
+
+    public String getBucketName(final Config systemConfig)
```
Comment (szyn) [inline digdag-standards/src/main/java/io/digdag/standards/command/ecs/TemporalProjectArchiveStorage.java:59]: I understand this method expects to get an S3 bucket name so I believe we can simplify this like
```suggestion
    public String getS3BucketName(final Config systemConfig)
    {
        return systemConfig.get(PARAMS_PREFIX + "s3.bucket", String.class);
    }    
```

**Review by szyn — COMMENTED** [review]
Sorry for the late reply.
I agree with adding `ecs.<name>.use_environment_file` to the system property. Actually, this property can also be set from the task config, but since the default is false and does not affect the behavior, it should be fine.

I have left some minor comments, I would appreciate it if you take a look at them when you get a chance.

File: digdag-standards/src/main/java/io/digdag/standards/command/EcsCommandExecutor.java [hunk 2]
```diff
@@ -822,6 +887,17 @@ private static String createStorageKey(final TaskRequest request, final String l
                 .toString();
     }
 
+    private static String createS3BucketARN(final String Bucket, final String fileKey)
```
[the remark this answers is not in the document]
Comment (rariyama) [reply digdag-standards/src/main/java/io/digdag/standards/command/EcsCommandExecutor.java]: Thanks for your review. I fixed on this commit https://github.com/treasure-data/digdag/pull/1762/commits/ab567e47a8a3c08cc954c67100cdf294e29dd95e.

File: digdag-standards/src/main/java/io/digdag/standards/command/EcsCommandExecutor.java [hunk 3]
```diff
@@ -822,6 +887,17 @@ private static String createStorageKey(final TaskRequest request, final String l
...
+    {
+        // S3 File ARN: "arn:aws:s3::{torage_bucket}/{env_file}"
```
[the remark this answers is not in the document]
Comment (rariyama) [reply digdag-standards/src/main/java/io/digdag/standards/command/EcsCommandExecutor.java]: I fixed on this commit https://github.com/treasure-data/digdag/pull/1762/commits/ab567e47a8a3c08cc954c67100cdf294e29dd95e.

File: digdag-standards/src/main/java/io/digdag/standards/command/EcsCommandExecutor.java [hunk 4]
```diff
@@ -822,6 +887,17 @@ private static String createStorageKey(final TaskRequest request, final String l
...
+        return new StringBuilder()
+                .append("arn:aws:s3:::")
+                .append(Bucket)
```
[the remark this answers is not in the document]
Comment (rariyama) [reply digdag-standards/src/main/java/io/digdag/standards/command/EcsCommandExecutor.java]: I fixed on this commit https://github.com/treasure-data/digdag/pull/1762/commits/ab567e47a8a3c08cc954c67100cdf294e29dd95e.

File: digdag-standards/src/main/java/io/digdag/standards/command/ecs/TemporalProjectArchiveStorage.java (hunk 1 above)
Comment (rariyama) [reply digdag-standards/src/main/java/io/digdag/standards/command/ecs/TemporalProjectArchiveStorage.java]: I see. The storage is expected to be S3 so I fixed type with it on this commit https://github.com/treasure-data/digdag/pull/1762/commits/02030bd322a591c0eecfb761622eec65a7ee316b.

**Comment by rariyama** [comment]
@szyn 
Thanks for your review. I fixed the problems you commented so please confirm it if you are available.

**Review by szyn — APPROVED** [review]
Looks good to me. Thank you for your contribution.

**Pull request merged**